src/main/java/com/denizenscript/depenizen/bukkit/commands/noteblockapi/NBSCommand.java

```diff
@@ -115,12 +135,28 @@ public static void autoExecute(ScriptEntry scriptEntry,
                 }
                 sp.setAutoDestroy(true);
                 sp.setPlaying(true);
+                if (scriptEntry.shouldWaitFor()) {
+                    waitingForSongs.put(sp, scriptEntry);
+                }
+                else {
+                   scriptEntry.setFinished(true);
+                }
             }
             case STOP -> {
                 for (PlayerTag p : targets) {
                     NoteBlockAPI.stopPlaying(p.getPlayerEntity());
+                    scriptEntry.setFinished(true);
                 }
             }
         }
     }
+
+    @EventHandler
+    public void onSongEnds(SongEndEvent event) {
+        SongPlayer sp = event.getSongPlayer();
+        if (waitingForSongs.containsKey(sp)) {
+            waitingForSongs.get(sp).setFinished(true);
+            waitingForSongs.remove(sp);
```

Comment:
`remove` returns the instance, and does nothing wrong if it's missing, so:
`ScriptEntry scriptEntry = waitingForSongs.remove(sp); if (scriptEntry != null) { scriptEntry.setFinished(true); }`

Minor change - in hypothetical performance-sensitive code this would be a 3x reduction in time spent (3 table lookups vs 1), but here it's just a minor bit of cleanliness